fsmonitor.c

```diff
@@ -50,17 +55,29 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
 	}
 	istate->fsmonitor_dirty = fsmonitor_dirty;
 
+	assert(istate->fsmonitor_dirty->bit_size <= istate->cache_nr);
+
 	trace_printf_key(&trace_fsmonitor, "read fsmonitor extension successful");
 	return 0;
 }
 
 void fill_fsmonitor_bitmap(struct index_state *istate)
 {
-	unsigned int i;
+	unsigned int i, j;
 	istate->fsmonitor_dirty = ewah_new();
-	for (i = 0; i < istate->cache_nr; i++)
-		if (!(istate->cache[i]->ce_flags & CE_FSMONITOR_VALID))
-			ewah_set(istate->fsmonitor_dirty, i);
+	for (i = j = 0; i < istate->cache_nr; i++) {
+		unsigned int flags = istate->cache[i]->ce_flags;
+		switch (flags & (CE_REMOVE | CE_FSMONITOR_VALID)) {
+		case 0:
+			ewah_set(istate->fsmonitor_dirty, j);
+			/* fallthru */
+		case CE_FSMONITOR_VALID:
+			j++;
+		default:
+			/* do not include entries being removed in fsmonitor_dirty */
+			break;
+		}
```

Comment:
Heh. Now that I read this, I actually prefer your version, I now find it way more readable...

If you want to keep the `j` way, maybe this?

```c
if (flags & CE_REMOVE)
    continue;
if (!(flags & CE_FSMONITOR_VALID))
    ewah_set(istate->fsmonitor_dirty, j);
j++;
```

Reply:
But now that I think of it even more, I don't think the `j` way would be more upstreamable. I'd think that the way you had it before would make more sense, maybe using `skipped` instead of `removed`?

Reply:
>  I'd think that the way you had it before would make more sense, maybe using `skipped` instead of `removed`?

I will switch this back and I think `skipped` is a better name than `removed`, thanks!